From: Dimitri John Ledkov <xnox@ubuntu.com>
Date: Thu, 20 Jul 2017 13:48:31 +0100
Subject: Set UseDomains to true, by default, on Ubuntu.

On Ubuntu, fallback DNS servers are disabled, therefore we do not leak queries
to a preset 3rd party by default. In resolved, dnssec is also disabled by
default, as too much of the internet is broken and using Ubuntu users to debug
the internet is not very productive - most of the time the end-user cannot fix
or know how to notify the site owners about the dnssec mistakes. Inherintally
the DHCP acquired DNS servers are therefore trusted, and are free to spoof
records. Not trusting DNS search domains, in such scenario, provides limited
security or privacy benefits. From user point of view, this also appears to be
a regression from previous Ubuntu releases which do trust DHCP acquired search
domains by default.

Therefore we are enabling UseDomains by default on Ubuntu.

Users may override this setting in the .network files by specifying
[DHCP|IPv6AcceptRA] UseDomains=no|route options.
---
 man/systemd.network.xml        | 6 +++---
 src/network/networkd-network.c | 2 ++
 2 files changed, 5 insertions(+), 3 deletions(-)

--- a/man/systemd.network.xml
+++ b/man/systemd.network.xml
@@ -333,8 +333,8 @@
           setting in the [DHCPv6] section, regardless of the presence of routers on the link, or
           what flags the routers pass. See <literal>IPv6AcceptRA=</literal>.</para>
 
-          <para>Furthermore, note that by default the domain name specified through DHCP is not used
-          for name resolution. See option <option>UseDomains=</option> below.</para>
+          <para>Furthermore, note that by default the domain name specified through DHCP, on Ubuntu,
+          are used for name resolution. See option <option>UseDomains=</option> below.</para>
 
           <para>See the [DHCPv4] or [DHCPv6] sections below for further configuration options for the
           DHCP client support.</para>
@@ -1893,7 +1893,7 @@
           to the effect of the <option>Domains=</option> setting. If set to <option>route</option>, the
           domain name received from the DHCP server will be used for routing DNS queries only, but not
           for searching, similar to the effect of the <option>Domains=</option> setting when the
-          argument is prefixed with <literal>~</literal>. Defaults to false.</para>
+          argument is prefixed with <literal>~</literal>. Defaults to true on Ubuntu.</para>
 
           <para>It is recommended to enable this option only on trusted networks, as setting this
           affects resolution of all hostnames, in particular of single-label names. It is generally
@@ -2378,7 +2378,7 @@
           the effect of the <option>Domains=</option> setting. If set to <literal>route</literal>, the domain name
           received via IPv6 RA will be used for routing DNS queries only, but not for searching, similar to the
           effect of the <option>Domains=</option> setting when the argument is prefixed with
-          <literal>~</literal>. Defaults to false.</para>
+          <literal>~</literal>. Defaults to true on Ubuntu.</para>
 
           <para>It is recommended to enable this option only on trusted networks, as setting this affects resolution
           of all hostnames, in particular of single-label names. It is generally safer to use the supplied domain
--- a/src/network/networkd-network.c
+++ b/src/network/networkd-network.c
@@ -403,6 +403,7 @@
                 .dhcp_use_gateway = -1,
                 .dhcp_send_hostname = true,
                 .dhcp_send_release = true,
+                .dhcp_use_domains = DHCP_USE_DOMAINS_YES,
                 .dhcp_route_metric = DHCP_ROUTE_METRIC,
                 .dhcp_client_identifier = _DHCP_CLIENT_ID_INVALID,
                 .dhcp_route_table = RT_TABLE_MAIN,
@@ -474,6 +475,7 @@
 
                 .ipv6_accept_ra = -1,
                 .ipv6_accept_ra_use_dns = true,
+                .ipv6_accept_ra_use_domains = DHCP_USE_DOMAINS_YES,
                 .ipv6_accept_ra_use_gateway = true,
                 .ipv6_accept_ra_use_route_prefix = true,
                 .ipv6_accept_ra_use_autonomous_prefix = true,
